pkg/pool-weighted/contracts/smart/IndexPool.sol

```diff
@@ -333,17 +333,82 @@ contract IndexPool is BaseWeightedPool, ReentrancyGuard {
 
+        // check if uninitialized token will be swapped INTO the pool
         if (minBalances[swapRequest.tokenIn] != 0) {
-            currentBalanceTokenIn = minBalances[swapRequest.tokenIn];
+            /* 
+                check if swap makes token become initialized
+            */
+            if (currentBalanceTokenIn.add(swapRequest.amount) >= minBalances[swapRequest.tokenIn]) {
+                // initiate new weight change this time with the final target weights (from reindex call)
+                // and set targetWeight of initialized token to zero
+                _setOriginalTargetWeight(currentBalanceTokenIn, swapRequest);
+                // use minimumBalance one last time to calc swap price
+                currentBalanceTokenIn = minBalances[swapRequest.tokenIn];
+                // set minimumBalance for initialized token to zero
+                minBalances[swapRequest.tokenIn] = 0;
+            } else {
+                currentBalanceTokenIn = minBalances[swapRequest.tokenIn];
+            }
         }
 
         return super.onSwap(swapRequest, currentBalanceTokenIn, currentBalanceTokenOut);
     }
 
-    /// @dev Calculates the time horizon for a rebasing based on max weight change.
-    /// @param tokens Array with addresses of tokens.
-    /// @param desiredWeights Array with desired weights of tokens. Must be in same order.
-    /// @return changeTime Time horizon for the rebalancing period
+    /// @dev Initiates the weight change to the original target weight of initialized token
+    /// @notice Also resets target weight of initialized token to zero within _startGradualWeightChange to save gas
+    /// @param swapRequest Swap params.
+    /// @param currentBalanceTokenIn Vault balance of token that is swapped into the pool.
+    function _setOriginalTargetWeight(uint256 currentBalanceTokenIn, SwapRequest memory swapRequest) private {
+        (IERC20[] memory tokens, , ) = getVault().getPoolTokens(getPoolId());
+        // get original reindex values by interpolating back
+        uint256[] memory originalReindexTargets;
+        {
+            (, , uint256[] memory endWeights, , uint256[] memory finalWeights) = getGradualWeightUpdateParams();
+            originalReindexTargets = IndexPoolUtils.normalizeInterpolated(endWeights, finalWeights);
+        }
+
+        // create fixedWeights (e.g. 0/0/0/0/1) for startWeights & endWeights
+        uint256[] memory fixedEndWeights = new uint256[](tokens.length);
+        uint256[] memory fixedStartWeights = new uint256[](tokens.length);
+        uint256[] memory newTokenTargetWeights = new uint256[](tokens.length);
+
+        for (uint256 i = 0; i < tokens.length; i++) {
+            bytes32 tokenState = _tokenState[tokens[i]];
+            // get final weights for reindex tokens as given per reindex call
+            uint256 finalTokenTargetWeight = tokenState.decodeUint32(_NEW_TOKEN_TARGET_WEIGHT_OFFSET).uncompress32();
```

Comment:
Should we set the `tokenState`->NEW_TOKEN_TOKEN_TARGET_WEIGHT value back to zero after going into the if statements so it won't trigger the next time `_setOriginalTargetWeight` gets triggered?

Reply:
The target weight of the initialized token is reset to zero in `_startGradualWeightChange` so taht we have to write to storage only once. Is that what you mean?